### PR TITLE
feat(Sink): support redis sink

### PR DESCRIPTION
**Pull request by xxhZs**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Support redis sink.  Data in RW will be saved in Redis in the form of string key-value pairs. By default, our key is the primary key, and the value includes all other values. Users can also input their desired format in string format. Therefore, even in the case of 'append only', we still need to provide the primary key.

Please refer to the README for specific instructions.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by st1page** [comment]
related to https://github.com/risingwavelabs/risingwave/issues/11995？ Should it be `FORMAT UPSERT ENCODE PLAIN`?

**Comment by xiangjinwu** [comment]
> related to #11995 ? Should it be `FORMAT UPSERT ENCODE PLAIN`?

It is indeed a new encoding but we may need a better name than `plain`:
* It may be configured by a template string, as the integration test shows `username:{username},event_timestamp{event_timestamp}` (btw missing `:` for `event_timestamp`, or intentional to mean user can do anything incl bugs?)
* The default key template is like `213:alice:2023-01-12`
* The default value template is like `[t,laptop,123.73]`

Is the default really useful? Maybe we shall require the user to provide templates when using this new encoding, or design the default carefully.

Talking about implementation, `to_text` is supposed to stay in sync with PostgreSQL's `typoutput` of each type, which is unfortunately **impure** and depends on a lot of implicit session configs (`TimeZone `, `DateStyle`, `IntervalStyle`, `extra_float_digits`, `bytea_output`, etc). Currently `to_text` does not take any extra argument because we do not support any of the configs above except `TimeZone`, which is currently handled by the caller before calling `Timestamptz::to_text`. So the question is: are the strings in sink expected to match the PostgreSQL `typoutput`?

File: src/connector/src/sink/redis.rs [hunk 1]
```diff
@@ -38,35 +90,296 @@ impl Sink for RedisSink {
...
-        todo!()
+        Ok(RedisSinkWriter::new(
+            self.config.clone(),
+            self.schema.clone(),
+            self.pk_indices.clone(),
+            self.is_append_only,
+        )?)
     }
 
     async fn validate(&self, _client: Option<ConnectorClient>) -> Result<()> {
```
Comment (hzxa21) [inline src/connector/src/sink/redis.rs:101]: I think we should validate the key_foramt and value_format here to make sure the user-provided format matches with our spec. For example, the field name matches with an existing column in the schema; the field name is wrapped by `{}`, etc.

File: src/connector/src/sink/redis.rs [hunk 2]
```diff
@@ -38,35 +90,296 @@ impl Sink for RedisSink {
...
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        is_append_only: bool,
+    ) -> Result<Self> {
+        let client = config.common.build_client()?;
+        let conn = Some(client.get_connection()?);
+        let pipe = redis::pipe();
+        let kv_formatter = match (config.common.key_format, config.common.value_format) {
+            (Some(key_format), Some(value_format)) => Some((key_format, value_format)),
+            _ => None,
```
Comment (hzxa21) [inline src/connector/src/sink/redis.rs:131]: This means user needs to specifiy both key_format and value_format. Specifying only one of them will not work and the default is used for both. If this is the intention, can we add a check in `validate` to make sure specifying only one format is an error?

**Comment by xiangjinwu** [comment]
The backend part of #11995 has been merged. Check this [link](https://github.com/risingwavelabs/risingwave/compare/xxh/redis-sink...review-11999-sink-redis?expand=1) as an example of how it could be applied to redis and the new encoding here.

File: src/connector/src/sink/redis.rs [hunk 3]
```diff
@@ -40,28 +108,302 @@ impl Sink for RedisSink {
...
+                check_string_format(value_format, &all_set)?;
+            }
+            (Some(key_format), None) => {
+                check_string_format(key_format, &pk_set)?;
+            }
+            (None, Some(value_format)) => {
+                check_string_format(value_format, &all_set)?;
+            }
+            (None, None) => {}
+        };
```
Comment (tabVersion) [inline src/connector/src/sink/redis.rs:153]: shall we just make `check_string_format` accepts `Option<String>`

File: src/connector/src/sink/redis.rs [hunk 4]
```diff
@@ -40,28 +108,302 @@ impl Sink for RedisSink {
...
+        self.pipe.query(&mut self.conn.as_mut().unwrap())?;
+        self.pipe.clear();
+        Ok(())
     }
 }
 
-pub struct RedisSinkWriter;
+impl FormattedSink for RedisSinkPayloadWriter {
+    type K = String;
+    type V = Vec<u8>;
```
Comment (tabVersion) [inline src/connector/src/sink/redis.rs:197]: IIRC, `ToRedisArg` accepts bytes for both key and value. any reason for limiting key to string?

File: src/connector/src/sink/redis.rs [hunk 5]
```diff
@@ -12,25 +12,93 @@
...
-pub struct RedisSink;
+pub struct RedisSink {
+    config: RedisConfig,
+    schema: Schema,
+    is_append_only: bool,
+    pk_indices: Vec<usize>,
+}
+
+fn check_string_format(format: &str, set: &HashSet<String>) -> Result<()> {
+    let re = Regex::new(r"\{([^}]*)\}").unwrap();
```
Comment (tabVersion) [inline src/connector/src/sink/redis.rs:71]: will it handle format containing escaped string? such as `'{\"a\": \{{col_a}\}}'`

Comment (xxhZs) [reply src/connector/src/sink/redis.rs:71]: Not yet, escaped characters are treated as strings, and {} can only have column names in it, not other characters.

File: src/connector/src/sink/redis.rs (hunk 4 above)
Comment (xxhZs) [reply src/connector/src/sink/redis.rs:197]: Inserting `k v` into redis in string format is only supported here.

File: src/connector/src/common.rs [hunk 6]
```diff
@@ -440,6 +441,23 @@ impl ClickHouseCommon {
     }
 }
 
+#[derive(Deserialize, Serialize, Debug, Clone)]
```
Comment (wenym1) [inline src/connector/src/common.rs:444]: If this config is only used in redis, we should put in in the file of redis sink. ditto for clickhouse sink.

File: src/connector/src/sink/encoder/template.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+// limitations under the License.
+
+use risingwave_common::catalog::Schema;
+use risingwave_common::row::Row;
+use risingwave_common::types::ToText;
+
+use super::{JsonEncoder, Result, RowEncoder};
+use crate::sink::SinkError;
+
+pub enum RedisFormatEncoder {
```
Comment (wenym1) [inline src/connector/src/sink/encoder/template.rs:22]: I think we don't need an extra enum for redis. The `RedisFormatterEncoder` has no special redis logic, and `SinkFormatterImpl` is already an enum.

File: src/connector/src/sink/encoder/template.rs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    fn col_indices(&self) -> Option<&[usize]> {
+        self.col_indices.as_ref().map(Vec::as_ref)
+    }
+
+    fn encode_cols(
+        &self,
+        row: impl Row,
+        col_indices: impl Iterator<Item = usize>,
+    ) -> Result<Self::Output> {
+        let mut s = self.template.to_string();
```
Comment (wenym1) [inline src/connector/src/sink/encoder/template.rs:91]: We may add a TODO to improve the performance. 

1. We don't need to handle all columns. Only the column on the template string should be used for replace.
2. the current implementation should process the template string for each row, but actually we only need to process and compile the template string as something like a regrex, and then apply the row value to the compiled one.

File: src/connector/src/sink/formatter/mod.rs [hunk 9]
```diff
@@ -55,11 +56,12 @@ macro_rules! tri {
     };
 }
 
-#[expect(clippy::enum_variant_names)]
 pub enum SinkFormatterImpl {
     AppendOnlyJson(AppendOnlyFormatter<JsonEncoder, JsonEncoder>),
     UpsertJson(UpsertFormatter<JsonEncoder, JsonEncoder>),
     DebeziumJson(DebeziumJsonFormatter),
+    RedisAppendOnly(AppendOnlyFormatter<RedisFormatEncoder, RedisFormatEncoder>),
```
Comment (wenym1) [inline src/connector/src/sink/formatter/mod.rs:63]:
```
AppendOnlyTemplate(AppendOnlyFormatter<TemplateEncoder, TemplateEncoder>)
UpsertTemplate(UpsertFormatter<TemplateEncoder, TemplateEncoder>)
```

File: src/connector/src/sink/redis.rs [hunk 10]
```diff
@@ -12,25 +12,95 @@
...
+pub struct RedisSink {
+    config: RedisConfig,
+    schema: Schema,
+    is_append_only: bool,
+    pk_indices: Vec<usize>,
+}
+
+fn check_string_format(format: &Option<String>, set: &HashSet<String>) -> Result<()> {
+    if let Some(format) = format {
+        let re = Regex::new(r"\{([^}]*)\}").unwrap();
```
Comment (wenym1) [inline src/connector/src/sink/redis.rs:72]: May add comment to explain the meaning of such regexp.

File: src/connector/src/sink/formatter/mod.rs [hunk 11]
```diff
@@ -107,6 +109,70 @@ impl SinkFormatterImpl {
             )))
         }
     }
+
+    pub fn new_with_redis(
```
Comment (wenym1) [inline src/connector/src/sink/formatter/mod.rs:113]: The syntax for create redis sink with key value template format may follow the new `create sink ... format XXX encode XXX` syntax.

If this PR is urgent, we can move the template row encoder to a separate PR, and in this PR we only use the json encoder for both key and value.

File: src/connector/src/sink/redis.rs [hunk 12]
```diff
@@ -40,28 +110,288 @@ impl Sink for RedisSink {
...
+
+    #[cfg(test)]
+    pub fn mock() -> Self {
+        let conn = None;
+        let pipe = redis::pipe();
+        Self { conn, pipe }
+    }
+
+    pub fn commit(&mut self) -> Result<()> {
+        self.pipe.query(&mut self.conn.as_mut().unwrap())?;
```
Comment (wenym1) [inline src/connector/src/sink/redis.rs:177]: May use the async version?
https://docs.rs/redis/latest/redis/struct.Pipeline.html#method.query_async

File: src/connector/src/sink/formatter/mod.rs (hunk 11 above)
Comment (xxhZs) [reply src/connector/src/sink/formatter/mod.rs:125]: Yes, I will create a new pull request to implement the new syntax. I think we can initially implement it in the current way, and then create a new pull request for modifications.

**Review by wenym1 — APPROVED** [review]

**Review by hzxa21 — APPROVED** [review]
